plugins/optimization-detective/detect.js

```diff
@@ -371,49 +494,73 @@ export default async function detect( {
 			boundingClientRect: elementIntersection.boundingClientRect,
 		};
 
-		urlMetric.elements.push( elementMetrics );
+		urlMetric.elements.push( elementData );
+		elementsByXPath.set( elementData.xpath, elementData );
 	}
 
 	if ( isDebug ) {
 		log( 'Current URL metric:', urlMetric );
 	}
 
-	// Yield to main before sending data to server to further break up task.
+	// Wait for the page to be hidden.
 	await new Promise( ( resolve ) => {
-		setTimeout( resolve, 0 );
-	} );
-
-	try {
-		const restUrl = new URL( restApiEndpoint );
-		restUrl.searchParams.append( 'slug', urlMetricSlug );
-		restUrl.searchParams.append( 'nonce', urlMetricNonce );
-		const response = await fetch( restUrl, {
-			method: 'POST',
-			headers: {
-				'Content-Type': 'application/json',
-				'X-WP-Nonce': restApiNonce,
+		win.addEventListener( 'pagehide', resolve, { once: true } );
+		win.addEventListener( 'pageswap', resolve, { once: true } );
+		doc.addEventListener(
+			'visibilitychange',
+			() => {
+				if ( document.visibilityState === 'hidden' ) {
+					// TODO: This will fire even when switching tabs.
+					resolve();
+				}
 			},
-			body: JSON.stringify( urlMetric ),
-		} );
-
-		if ( response.status === 200 ) {
-			setStorageLock( getCurrentTime() );
-		}
+			{ once: true }
+		);
+	} );
 
-		if ( isDebug ) {
-			const body = await response.json();
-			if ( response.status === 200 ) {
-				log( 'Response:', body );
-			} else {
-				error( 'Failure:', body );
+	if ( extensions.size > 0 ) {
+		for ( const [
+			extensionModuleUrl,
+			extension,
+		] of extensions.entries() ) {
+			if ( extension.finalize instanceof Function ) {
+				try {
+					await extension.finalize( {
+						isDebug,
+						getRootData,
+						getElementData,
+						amendElementData,
+						amendRootData,
+					} );
+				} catch ( err ) {
+					error(
+						`Unable to finalize module '${ extensionModuleUrl }':`,
+						err
+					);
+				}
 			}
 		}
-	} catch ( err ) {
-		if ( isDebug ) {
-			error( err );
-		}
 	}
 
+	// Even though the server may reject the REST API request, we still have to set the storage lock
+	// because we can't look at the response when sending a beacon.
+	setStorageLock( getCurrentTime() );
+
+	if ( isDebug ) {
+		log( 'Sending URL metric:', urlMetric );
+	}
+
+	const url = new URL( restApiEndpoint );
+	url.searchParams.set( '_wpnonce', restApiNonce );
+	url.searchParams.set( 'slug', urlMetricSlug );
+	url.searchParams.set( 'nonce', urlMetricNonce );
```

Comment:
One nonce is just for access to the REST API. 

The other nonce is to authorize the specific the action of storing the URL metric. In reality, the nonce here is serving as an [HMAC](https://en.wikipedia.org/wiki/HMAC). It is guaranteeing that the URL being submitted and the associated query vars slug hash are valid.

Reply:
Maybe this should be renamed from `nonce` to `hmac` but that should probably be done in another PR since the `nonce` param being set here is just a refactor from what is in `trunk` now. It's just that before it was setting it in the JSON body, whereas now it is being set as a GET param.